### PR TITLE
Implement extension loading

### PR DESCRIPTION
**Pull request by scolby33**
Re issue #81

### CODE REVIEW

**Comment by scolby33** [comment]
This definitely needs some work. Use cases we want to support:

```python
from pybel.ext import extension
from pybel.ext.extension import member
```

The former I've already done.

The latter might (almost certainly will) require messing with Python's import machinery.

**Comment by scolby33** [comment]
I've added some tests, but I'm not sure they can be proven to actually test what we want them to. Because of the `sys.modules` cache, they might not be as independent as we'd like. Should I start deleting things from the cache before each test? Should I be deleting things out of globals?

File: tests/test_extensions.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import unittest
+
+
+class TestExtensions(unittest.TestCase):
+    def test_import_extension(self):
+        import pybel.ext.test
+
+        assert pybel.ext.test.an_extension_function() == 42
```
Comment (cthoyt) [inline tests/test_extensions.py]: umm what about Yoda?

File: tox.ini [hunk 2]
```diff
@@ -43,4 +44,4 @@ deps = coverage
 skip_install = true
 commands =
     coverage combine
-    coverage report
\ No newline at end of file
+    coverage report
```
Comment (cthoyt) [inline tox.ini]: shame on me

File: src/pybel/ext/__init__.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+__all__ = [entry_point.name for entry_point in iter_entry_points(group=GROUP_NAME, name=None)]
+
+
+def setup():
+    from ..exthook import ExtensionImporter
+    importer = ExtensionImporter(GROUP_NAME)
+    importer.install()
+
+setup()
+del setup
```
Comment (cthoyt) [inline src/pybel/ext/__init__.py]: def janky

File: tests/test_extensions.py (hunk 1 above)
Comment (scolby33) [inline tests/test_extensions.py]: What?

File: src/pybel/ext/__init__.py (hunk 3 above)
Comment (scolby33) [inline src/pybel/ext/__init__.py]: I'm not sure why it's important to `del` it other than to keep the namespace clean. I just copied flask here, to be honest.

Comment (cthoyt) [inline src/pybel/ext/__init__.py]: I saw that's what they did in Flask, too. This is fine, but we should still call it what it is

File: tests/test_extensions.py (hunk 1 above)
Comment (cthoyt) [inline tests/test_extensions.py]: `assert 42 == pybel.ext.test.an_extension_function()` or bust

Comment (scolby33) [inline tests/test_extensions.py]: But...why? It reads much more naturally to me the first way.

Comment (cthoyt) [inline tests/test_extensions.py]: Unqualified disagree

**Comment by cthoyt** [comment]
Docs, then lets merge

File: src/pybel/ext/__init__.py [hunk 4]
```diff
@@ -1,14 +1,47 @@
...
+
+This works just like the standard :code:`console_scripts` entry point and the syntax follows all the same rules.
+
+Your extension will then be importable as :code:`pybel.ext.name_of_your_extension`
+
+:code:`pybel.ext` exists as a package on its own to trick the Python import system...if we just did this work in PyBEL's :code:`__init__.py`, there could be trouble.
+"""
 from pkg_resources import iter_entry_points
 
 GROUP_NAME = 'pybel.ext'
```
Comment (scolby33) [inline src/pybel/ext/__init__.py]: Should I do something clever with this line involving `__name__` or something, or just leave it hardcoded as `'pybel.ext'`? If the former, what would the solution be?

**Comment by scolby33** [comment]
It says I changed a lot more files than I did. I _may_ have screwed up the pull from develop...

**Comment by cthoyt** [comment]
Closes #81

**Pull request merged**